### ISSUE TITLE
All files export-ignore'd in gitattributes are marked a new files

### ISSUE THREAD

**Issue by gerardroche** [jisaacks/GitGutter#409]
*This is probably the same issue causing #388 (All lines of all existing files shows as newly inserted).*

I've had this issue for ages but I couldn't reproduce it. Some files were just marked as new. So I actually disabled that feature and just set `show_markers_on_untracked_file` to false so the issue was just hidden from me.

I got frustrated by it again because I needed to the changes in the current file, finally debugged it, and then it dawned on me what the issue is:

All files that are ignore'd on export by `.gitattributes` are marked as new (inserted) by GitGutter because GitGutter thinks they are new files. It thinks they're new because internally GitGutter uses `git archive` to export the current file to do a diff against it, but if it's ignore'd on export then `git archive` doesn't export it and so GitGutter thinks it is a new file.

Example

`.gitattributes`

```
# Files and directories with the attribute export-ignore won’t be added to
# archive files. See http://git-scm.com/docs/gitattributes for details.
/tests/             export-ignore
```

All files in the tests folder will be marked as new files.

I thought that git archive would have an option to ignore rules defined in `.gitattributes` file but surprisingly it doesn't. See [git archive docs](https://git-scm.com/docs/git-archive).

**Comment by deathaxe** [comment]
GitGutter uses `git archive` to read a file's content from git's database as this is the only known way to apply smudge filters correctly. This change was made to fix Issue #74. 

So any file marked as `export-ignore` will not be visible to GitGutter at all by nature. This could be fixed only by an alternative way of reading file content from git database with smudge filters being applied. The only known methods supportung smudge are `git archive`, `git diff`, `git checkout`. As checkout would overwrite the existing working file on disk, it can't be used.

**Comment by gerardroche** [comment]
I think this is a really big issue. Lots of projects use export-ignore. I've never used filters, not sure if I've ever worked on a project that did. Certainly far more projects use export-ignore than use filters. And the fix for the filters doesn't even fix that issue properly. It looks like it adds a performance overhead too. 

GitGutter should use export-ignore it to exclude `release_messages/`, `tests/`, and other non-production files from releases.

Many affected by this must see it as just one of those bugs that get eventually gets fixed. They've probably come to the issue tracker like I did and read all kinds of similar issues and tried reinstalling, setting the binary path, maybe it's a line ending issue, or whitespace issue. I went through those issues and ended up `show_markers_on_untracked_file` to false which just completely hides the problem. I assumed it was just some files. It was only later when I decided to debug it I realised it was export-ignore and then realised that most project tests/ directories are being marked as new by GitGutter.

I'd prefer to do without filters support. I don't think will affect me as I don't use them.

**Comment by deathaxe** [comment]
> And the fix for the filters doesn't even fix that issue properly. 

This answer is too simple and therefore wrong. Smudge filters can be used to do any stupid stuff during checkout or commit, but there are technical limitations as GitGutter expects the checked out file to be located in %temp% directory and the content from view is not handled by git at all.

> It looks like it adds a performance overhead too.

Can't confirm this assertion. On the one hand several benchmarks made to compare `git archive` and `git show` calls were performed with same result. On the other hand files are read from git repository only once after opening a file and once after each "Compare against ..." call. Thus there is no performance impact during normal work.

> GitGutter should use export-ignore it to exclude release_messages/, tests/, and other non-production files from releases.

GitGutter can't exclude `release_messages/` as this folder contains the messages being displayed in ST after upgrading the packages. Parts of it may be worth consideration.

> I'd prefer to do without filters support. I don't think it will affect me as I don't use them.

The feedback on fix for issue #74 proved there are several people who work with filters, even though they need to follow some rules to make them work with GitGutter properly. Thus I won't reintroduce a bug in favor of a fix for another one because of a single's persons meaning.

I'll accept a fix which addresses both issues, only!

>  It was only later when I decided to debug it I realized it was export-ignore

Very much thanks for taking the time to debug this issue and find the reason for. I haven't been aware of this feature at all as I never needed to use it. The knowledge about it may help to find a solution somehow. Already have an idea.
